chia/full_node/mempool.py

```diff
@@ -248,23 +257,68 @@ def remove_from_pool(self, items: List[bytes32], reason: MempoolRemoveReason) ->
                 self.fee_estimator.remove_mempool_item(info, iteminfo)
 
-    def add_to_pool(self, item: MempoolItem) -> None:
+    def add_to_pool(self, item: MempoolItem) -> Optional[Err]:
         """
         Adds an item to the mempool by kicking out transactions (if it doesn't fit), in order of increasing fee per cost
         """
 
         assert item.fee < MEMPOOL_ITEM_FEE_LIMIT
         assert item.npc_result.conds is not None
+        assert item.cost <= self.mempool_info.max_block_clvm_cost
 
-        # TODO: this block could be simplified by removing all items in a single
-        # SQL query. Or at least figure out which items to remove and then
-        # remove them all in a single call to remove_from_pool()
         with self._db_conn:
-            while self.at_full_capacity(item.cost):
-                # pick the item with the lowest fee per cost to remove
-                cursor = self._db_conn.execute("SELECT name FROM tx ORDER BY fee_per_cost ASC, seq DESC LIMIT 1")
-                name = bytes32(cursor.fetchone()[0])
-                self.remove_from_pool([name], MempoolRemoveReason.POOL_FULL)
+            # we have certain limits on transactions that will expire soon
+            # (in the next 15 minutes)
+            block_cutoff = self._block_height + 48
+            time_cutoff = self._timestamp + 900
+            if (item.assert_before_height is not None and item.assert_before_height < block_cutoff) or (
+                item.assert_before_seconds is not None and item.assert_before_seconds < time_cutoff
+            ):
+                # this lists only transactions that expire soon, in order of
+                # lowest fee rate along with the cumulative cost of such
+                # transactions counting from highest to lowest fee rate
+                cursor = self._db_conn.execute(
+                    """
+                    SELECT name,
+                        fee_per_cost,
+                        SUM(cost) OVER (ORDER BY fee_per_cost DESC, seq ASC) AS cumulative_cost
+                    FROM tx
+                    WHERE assert_before_seconds IS NOT NULL OR assert_before_height IS NOT NULL
+                    ORDER BY cumulative_cost DESC
+                    """,
+                )
+                to_remove: List[bytes32] = []
+                for row in cursor:
+                    name, fee_per_cost, cumulative_cost = row
+
+                    # there's space for us, stop pruning
+                    if cumulative_cost + item.cost <= self.mempool_info.max_block_clvm_cost:
+                        break
+
+                    # we can't evict any more transactions, abort (and don't
+                    # evict what we put aside in "to_remove".
+                    if fee_per_cost > item.fee_per_cost:
+                        return Err.INVALID_FEE_LOW_FEE
+                    to_remove.append(name)
+                self.remove_from_pool(to_remove, MempoolRemoveReason.POOL_FULL)
```

Comment:
@aqk does this removal reason make sense? the mempool isn't full, it's just using a stricter limit for a subset of transactions